GCS DeleteBucket expects the bucket to be empty to be deleted

### Expected Behavior

The given bucket should be deleted. We can provide an option to delete non-empty bucket, or explicitly mention that only empty buckets will be deleted.

### Actual Behaviour

The non-empty bucket does not get deleted. The task fails with the following error:

```sh
2024-01-29 16:09:11.283Deleting bucket 'smantri-gcs-kestra-test-bucket'
2024-01-29 16:09:12.910The bucket you tried to delete is not empty.
2024-01-29 16:09:12.910409 Conflict
DELETE https://storage.googleapis.com/storage/v1/b/smantri-gcs-kestra-test-bucket
{
  "code" : 409,
  "errors" : [ {
    "domain" : "global",
    "message" : "The bucket you tried to delete is not empty.",
    "reason" : "conflict"
  } ],
  "message" : "The bucket you tried to delete is not empty."
}
2024-01-29 16:09:12.910com.google.cloud.storage.StorageException: The bucket you tried to delete is not empty.
	at com.google.cloud.storage.StorageException.translate(StorageException.java:165)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:313)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.delete(HttpStorageRpc.java:642)
	at com.google.cloud.storage.StorageImpl.lambda$delete$14(StorageImpl.java:516)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.Retrying.run(Retrying.java:65)
	at com.google.cloud.storage.StorageImpl.run(StorageImpl.java:1515)
	at com.google.cloud.storage.StorageImpl.delete(StorageImpl.java:516)
	at io.kestra.plugin.gcp.gcs.DeleteBucket.run(DeleteBucket.java:54)
	at io.kestra.plugin.gcp.gcs.DeleteBucket.run(DeleteBucket.java:19)
	at io.kestra.core.runners.Worker$WorkerThread.run(Worker.java:684)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 409 Conflict
DELETE https://storage.googleapis.com/storage/v1/b/smantri-gcs-kestra-test-bucket
{
  "code" : 409,
  "errors" : [ {
    "domain" : "global",
    "message" : "The bucket you tried to delete is not empty.",
    "reason" : "conflict"
  } ],
  "message" : "The bucket you tried to delete is not empty."
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$3.interceptResponse(AbstractGoogleClientRequest.java:466)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:552)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:493)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:603)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.delete(HttpStorageRpc.java:638)
	... 10 more
```

### Steps To Reproduce

in the DeleteBucket task, provide a non0-empty bucket to be deleted.

### Environment Information

- Kestra Version: 0.13.8
- Plugin version: 0.13.8
- Operating System (OS / Docker / Kubernetes): Docker
- Java Version (If not docker):


### Example flow

```yaml
id: gcp-gcs-delete-bucket
namespace: company.team
tasks:
  - id: "delete"
    type: "io.kestra.plugin.gcp.gcs.DeleteBucket"
    serviceAccount: "{{ secret('GCP_SERVICE_ACCOUNT_JSON') }}"
    projectId: "<project-name>"
    name: "<bucket-name>"
```

--- a/src/main/java/io/kestra/plugin/gcp/gcs/DeleteBucket.java
+++ b/src/main/java/io/kestra/plugin/gcp/gcs/DeleteBucket.java
@@ -2,18 +2,18 @@ package io.kestra.plugin.gcp.gcs;
 
 import com.google.cloud.storage.Storage;
 import com.google.cloud.storage.StorageException;
-import io.kestra.core.models.property.Property;
-import io.swagger.v3.oas.annotations.media.Schema;
-import lombok.*;
-import lombok.experimental.SuperBuilder;
 import io.kestra.core.models.annotations.Example;
 import io.kestra.core.models.annotations.Plugin;
+import io.kestra.core.models.property.Property;
 import io.kestra.core.models.tasks.RunnableTask;
 import io.kestra.core.runners.RunContext;
+import io.swagger.v3.oas.annotations.media.Schema;
+import jakarta.validation.constraints.NotNull;
+import lombok.*;
+import lombok.experimental.SuperBuilder;
 import org.slf4j.Logger;
 
 import java.net.URI;
-import jakarta.validation.constraints.NotNull;
 
 @SuperBuilder
 @ToString
@@ -53,6 +53,8 @@ public class DeleteBucket extends AbstractGcs implements RunnableTask<DeleteBuck
     @Builder.Default
     private Property<Boolean> force = Property.ofValue(false);
 
+    public static final int CONCURRENT_DELETIONS = 8;
+
     @Override
     public Output run(RunContext runContext) throws Exception {
         Storage connection = this.connection(runContext);
@@ -69,7 +71,7 @@ public class DeleteBucket extends AbstractGcs implements RunnableTask<DeleteBuck
                 .serviceAccount(this.serviceAccount)
                 .scopes(this.scopes)
                 .from(Property.ofValue("gs://" + name))
-                .concurrent(8)
+                .concurrent(CONCURRENT_DELETIONS)
                 .build();
 
             deleteListTask.run(runContext);

--- a/src/main/java/io/kestra/plugin/gcp/gcs/DeleteList.java
+++ b/src/main/java/io/kestra/plugin/gcp/gcs/DeleteList.java
@@ -75,7 +75,7 @@ public class DeleteList extends AbstractList implements RunnableTask<DeleteList.
 
     @Min(2)
     @Schema(
-        title = "Number of concurrent parallel deletions"
+        title = "Number of concurrent deletions"
     )
     @PluginProperty
     private Integer concurrent;

--- a/src/test/java/io/kestra/plugin/gcp/gcs/DeleteBucketTest.java
+++ b/src/test/java/io/kestra/plugin/gcp/gcs/DeleteBucketTest.java
@@ -1,0 +1,118 @@
+package io.kestra.plugin.gcp.gcs;
+
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import com.google.common.collect.ImmutableMap;
+import io.kestra.core.junit.annotations.KestraTest;
+import io.kestra.core.models.property.Property;
+import io.kestra.core.runners.RunContext;
+import io.kestra.core.runners.RunContextFactory;
+import io.kestra.core.utils.TestsUtils;
+import jakarta.inject.Inject;
+import org.junit.jupiter.api.Test;
+import org.mockito.MockedConstruction;
+import org.mockito.Mockito;
+
+import java.net.URI;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@KestraTest
+class DeleteBucketTest {
+    @Inject
+    private RunContextFactory runContextFactory;
+
+    @Test
+    void forceFalse_shouldDeleteBucket() throws Exception {
+        // given
+        DeleteBucket task = Mockito.spy(DeleteBucket.builder()
+            .id(DeleteBucket.class.getSimpleName())
+            .type(DeleteBucket.class.getName())
+            .name(Property.ofValue("my-bucket"))
+            .force(Property.ofValue(false))
+            .build());
+
+        // Mock GCS Storage delete so no real bucket is touched.
+        Storage storage = mock(Storage.class);
+        doReturn(storage).when(task).connection(any(RunContext.class));
+        when(storage.delete("my-bucket")).thenReturn(true);
+
+        // Real (non-mocked) Kestra RunContext
+        var runContext = TestsUtils.mockRunContext(this.runContextFactory, task, ImmutableMap.of());
+
+        // when
+        DeleteBucket.Output output = task.run(runContext);
+
+        // then
+        verify(storage, times(1)).delete("my-bucket");
+        assertThat(output.getBucket(), is("my-bucket"));
+        assertThat(output.getBucketUri(), is(new URI("gs://my-bucket")));
+    }
+
+    @Test
+    void forceTrue_shouldCallDeleteListBeforeDeletingBucket() throws Exception {
+        // given
+        DeleteBucket task = Mockito.spy(DeleteBucket.builder()
+            .id(DeleteBucket.class.getSimpleName())
+            .type(DeleteBucket.class.getName())
+            .name(Property.ofValue("my-bucket"))
+            .force(Property.ofValue(true))
+            .build());
+
+        Storage storage = mock(Storage.class);
+        doReturn(storage).when(task).connection(any(RunContext.class));
+        when(storage.delete("my-bucket")).thenReturn(true);
+
+        var runContext = TestsUtils.mockRunContext(this.runContextFactory, task, ImmutableMap.of());
+
+        // Intercept DeleteList construction so we don't delete anything real.
+        try (MockedConstruction<DeleteList> mocked = Mockito.mockConstruction(
+            DeleteList.class,
+            (mock, context) -> {
+                when(mock.run(any(RunContext.class)))
+                    .thenReturn(DeleteList.Output.builder().count(0).size(0).build());
+            }
+        )) {
+            // when
+            DeleteBucket.Output output = task.run(runContext);
+
+            // then: DeleteList constructed once and executed once
+            assertThat(mocked.constructed().size(), is(1));
+            verify(mocked.constructed().getFirst(), times(1)).run(runContext);
+
+            // and: bucket delete executed
+            verify(storage, times(1)).delete("my-bucket");
+
+            // and: output ok
+            assertThat(output.getBucket(), is("my-bucket"));
+            assertThat(output.getBucketUri(), is(new URI("gs://my-bucket")));
+        }
+    }
+
+    @Test
+    void shouldThrow404WhenBucketNotFound() throws Exception {
+        // given
+        DeleteBucket task = Mockito.spy(DeleteBucket.builder()
+            .id(DeleteBucket.class.getSimpleName())
+            .type(DeleteBucket.class.getName())
+            .name(Property.ofValue("missing-bucket"))
+            .force(Property.ofValue(false))
+            .build());
+
+        Storage storage = mock(Storage.class);
+        doReturn(storage).when(task).connection(any(RunContext.class));
+        when(storage.delete("missing-bucket")).thenReturn(false);
+
+        var runContext = TestsUtils.mockRunContext(this.runContextFactory, task, ImmutableMap.of());
+
+        // when / then
+        StorageException ex = assertThrows(StorageException.class, () -> task.run(runContext));
+        assertThat(ex.getCode(), is(404));
+        assertThat(ex.getMessage(), containsString("Couldn't find bucket 'missing-bucket'"));
+    }
+}

